src/zap/zapinfo.cpp

```diff
@@ -2136,19 +2136,55 @@ DWORD FilterMethodAttribsForIsSupported(DWORD attribs, CORINFO_METHOD_HANDLE ftn
-        if (isX86intrinsic
-            || strcmp(namespaceName, "System.Runtime.Intrinsics.Arm.Arm64") == 0
-            || strcmp(namespaceName, "System.Runtime.Intrinsics") == 0)
+        // If it's an intrinsic that requires VEX encoding, do not report as intrinsic
+        // to force this to become a regular method call.
+        // We don't allow RyuJIT to use VEX encoding at AOT compilation time, so these
+        // cannot be pregenerated. Not reporting them as intrinsic will make sure
+        // it will do the right thing at runtime (the called method will be JITted).
+        // It will be slower, but correct.
+        if (fIsX86intrinsic
+            && (
+                strcmp(className, "Avx") == 0 || strcmp(className, "Fma") == 0 || strcmp(className, "Avx2") == 0 || strcmp(className, "Bmi1") == 0 || strcmp(className, "Bmi2") == 0
+                || (
+                    strcmp(className, "X64") == 0
+                    && (
+                        strcmp(enclosingClassName, "Bmi1") == 0 || strcmp(enclosingClassName, "Bmi2") == 0
+                        )
+                    )
+                )
+            )
         {
-            attribs &= ~CORINFO_FLG_JIT_INTRINSIC;
+            // We do want the IsSupported for VEX instructions to be recognized as intrinsic so that the
+            // potentially worse quality code doesn't actually run until tiered JIT starts
+            // kicking in and recompiling methods. Reporting this as intrinsic makes RyuJIT expand it
+            // into `return false`.
+            if (fIsGetIsSupportedMethod)
+                return attribs;
+
+            // Treat as a regular method call (into a JITted method).
+            return (attribs & ~CORINFO_FLG_JIT_INTRINSIC) | CORINFO_FLG_DONT_INLINE;
+        }
 
-            // The intrinsic just recursively calls itself in IL so it wouldn't inline anyway,
-            // but let's make that explicit.
-            attribs |= CORINFO_FLG_DONT_INLINE;
+#endif // defined(_TARGET_X86_) || defined(_TARGET_AMD64_)
+
+        // Do not report the get_IsSupported method as an intrinsic. This will turn the call into a regular
+        // call. We also make sure none of the hardware intrinsic method bodies get pregenerated in crossgen
+        // (see ZapInfo::CompileMethod) but get JITted instead. The JITted method will have the correct
+        // answer for the CPU the code is running on.
+        if (fIsGetIsSupportedMethod &&
+#if defined(_TARGET_X86_) || defined(_TARGET_AMD64_)
+            fIsX86intrinsic ||
+#elif _TARGET_ARM64_
+            strcmp(namespaceName, "System.Runtime.Intrinsics.Arm.Arm64") == 0 ||
+#endif
+            strcmp(namespaceName, "System.Runtime.Intrinsics") == 0)
+        {
+            // Treat as a regular method call (into a JITted method).
+            return (attribs & ~CORINFO_FLG_JIT_INTRINSIC) | CORINFO_FLG_DONT_INLINE;
```

Comment:
Same concern as I called out above. Intrinsics are special and are handled in the importer before inlining checks and the recursive calls are dependent on `mustExpand` (which itself depends on `JIT_INTRINSIC` flag).

It might be better to just fixup the `compiler::impIntrinsic` method to have a `if (isCompilingForAOT)` path for the `IsSupported` checks, which returns `nullptr` rather than a `gtNewIconNode` (see https://github.com/dotnet/coreclr/blob/master/src/jit/importer.cpp#L3486-L3494, as we handle the `IsSupported` and `PlatformNotSupportedException` intrinsics a bit earlier).

Reply:
That makes sense to me, though note that the JIT doesn't know whether it's SPC.dll or not.

Reply:
Is that actually a concern now that we have a mechanism to handle functions which consume AVX/AVX2/FMA/BMI1/BMI2 methods, so things work even if they are called without an `IsSupported` check?

I believe the "worst case" scenario is someone doesn't use the check and calls, for example `Sse41.DotProduct` on a processor that doesn't support it. We will have emitted the correct instruction encoding and that will trigger a hardware exception as the instruction isn't supported.

Reply:
There is also the chance for subtle bugs, if users don't provide software fallbacks, but I think it might  be better to just have that be upfront in the initial release. If we ever want to make this more generally available, users will always have that scenario.

Reply:
Right - I don't think it is a concern that we don't specially handle SPC.dll here.

> I believe the "worst case" scenario is someone doesn't use the check and calls, for example Sse41.DotProduct on a processor that doesn't support it.

But the code won't work in any event. I guess if they were actually relying on getting PNSE it might. I suppose we could trap the hardware exception and throw PNSE, but I'm not sure that complication is justified.

Reply:
I agree, I think it would be better to encourage people to have the appropriate `IsSupported` checks 😄